File: apps/mobile_ios/test/metadata_registry_test.dart
// Guard-rail: every `runs.metadata` key used anywhere in the mobile-side
// Dart codebase must be registered in `docs/backend/metadata.md`.
//
// `runs.metadata` is a jsonb bag with no type-level protection. Cross-client
// drift (mobile writes `activity_type`, web reads `activityType`) is the
// exact failure mode the registry was created to prevent. This test keeps
// the registry and the code from diverging silently — if you add a key in
// Dart without adding a row to `docs/backend/metadata.md`, CI fails here.
//
// Scope: Dart-only. Equivalent tests should exist for TypeScript (web),
// Kotlin (watch_wear), and Swift (watch_ios) — tracked as a TODO so the
// cross-platform story is symmetrical. Today the registry itself is the
// coordination point for those platforms.

import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// Keys that are written by the generator itself, so adding them would be
/// circular. Keep this list empty unless there's a genuine reason to
/// exclude a key from the discipline.
const _exemptKeys = <String>{};

/// Heuristics that pick up non-key matches we want to ignore. For example:
/// some helper extracts keys *from* `metadata` in a loop using
/// `metadata[entry.key]` — `entry` isn't a string literal and the regex
/// already skips that case. Keep this empty until proven necessary.
const _exemptReferences = <String>{};

void main() {
  test(
    'every `runs.metadata` key referenced in Dart source is registered in docs/backend/metadata.md',
    () {
      final registry = _parseRegistry(
        File('../../docs/backend/metadata.md').readAsStringSync(),
      );
      final constMap = _parseMetadataKeyConstants();

      final roots = <Directory>[
        Directory('lib'),
        Directory('../../packages/api_client/lib'),
        Directory('../../packages/run_recorder/lib'),
      ].where((d) => d.existsSync());

      final referenced = <String, List<String>>{};
      for (final root in roots) {
        for (final entity in root.listSync(recursive: true)) {
          if (entity is! File) continue;
          if (!entity.path.endsWith('.dart')) continue;
          if (entity.path.contains('/generated/')) continue;
          final src = entity.readAsStringSync();
          for (final k in _extractMetadataKeys(src, constMap)) {
            referenced.putIfAbsent(k, () => []).add(entity.path);
          }
        }
      }

      final unknown = referenced.keys.toSet()
        ..removeAll(registry)
        ..removeAll(_exemptKeys)
        ..removeAll(_exemptReferences);

      if (unknown.isNotEmpty) {
        final lines = <String>[
          'Unregistered metadata key(s) referenced in Dart source:',
          '',
          for (final k in unknown.toList()..sort()) ...[
            '  $k',
            ...referenced[k]!.map((p) => '      at $p'),
          ],
          '',
          'Either:',
          '  1) Register the key in docs/backend/metadata.md (preferred) — use '
              'snake_case, describe the shape, writers, readers. See the '
              'top of that doc for the rule.',
          '  2) If the match is spurious (pattern matched a non-metadata '
              "reference), add the key to _exemptReferences in this test "
              'with a one-line comment explaining why.',
          '',
          'Drift in `runs.metadata` is invisible to the DB type system '
              '— this registry IS the coordination point across mobile, '
              'web, watch_wear, and watch_ios.',
        ];
        fail(lines.join('\n'));
      }
    },
  );

  test(
    'every registered metadata key is referenced in Dart source (dead-key check)',
    () {
      final registry = _parseRegistry(
        File('../../docs/backend/metadata.md').readAsStringSync(),
      );
      final constMap = _parseMetadataKeyConstants();

      final referenced = <String>{};
      final roots = <Directory>[
        Directory('lib'),
        Directory('../../packages/api_client/lib'),
        Directory('../../packages/run_recorder/lib'),
      ].where((d) => d.existsSync());
      for (final root in roots) {
        for (final entity in root.listSync(recursive: true)) {
          if (entity is! File) continue;
          if (!entity.path.endsWith('.dart')) continue;
          if (entity.path.contains('/generated/')) continue;
          referenced
              .addAll(_extractMetadataKeys(entity.readAsStringSync(), constMap));
        }
      }

      // A registered key that no Dart client references may still be
      // legitimately written by the web / watches / an Edge Function. So
      // we only flag keys declared in `docs/backend/metadata.md` as being written
      // BY a mobile-side writer but not actually referenced in Dart source
      // — a true dead key.
      //
      // Since the registry doesn't mark "Dart-written" separately from
      // "any client writes this", this test errs on the side of caution
      // and only warns in the reason string. To make it a hard failure,
      // scope by writer in a follow-up.
      final unused = registry.difference(referenced);
      // Intentionally NOT failing on unused — just surfacing so the next
      // session has a starting list. Converted to a hard failure once the
      // writer column in docs/backend/metadata.md is machine-readable.
      if (unused.isNotEmpty) {
        // Intentional: printed, not failed. Upgrade to `fail(...)` once
        // the writer column is structured enough to exclude non-Dart-only
        // keys automatically.
        // ignore: avoid_print
        print(
          'INFO: these registered keys are not referenced in Dart source '
          '(may be web/watch/EF-only): ${(unused.toList()..sort()).join(', ')}',
        );
      }
    },
  );
}

/// Extract registered key names from the markdown registry. Each row in
/// each table starts with `| \`key\` |`. Match the first backticked token.
Set<String> _parseRegistry(String doc) {
  final out = <String>{};
  final re = RegExp(r'^\|\s*`([a-z_][a-z0-9_]*)`\s*\|', multiLine: true);
  for (final m in re.allMatches(doc)) {
    out.add(m.group(1)!);
  }
  return out;
}

/// Parse `MetadataKeys` constant declarations from `core_models` into an
/// `identifier -> wire value` map (e.g. `activityType -> activity_type`).
/// The keys are increasingly centralised behind `MetadataKeys.*` instead of
/// raw string literals, so the scanner must resolve those references back to
/// the wire value or it would silently lose coverage of every centralised
/// key.
Map<String, String> _parseMetadataKeyConstants() {
  final file = File(
    '../../packages/core_models/lib/src/metadata_keys.dart',
  );
  if (!file.existsSync()) return const {};
  final src = file.readAsStringSync();
  final out = <String, String>{};
  final re = RegExp(
    r"""static\s+const\s+String\s+([A-Za-z0-9_]+)\s*=\s*'([a-z_][a-z0-9_]*)'""",
  );
  for (final m in re.allMatches(src)) {
    out[m.group(1)!] = m.group(2)!;
  }
  return out;
}

/// Find every metadata-key reference in a single Dart source file. Covers:
///
///   * subscript access:  `metadata['xxx']`, `metadata!['xxx']`,
///                        `metadata?['xxx']`, `run.metadata!['xxx']`,
///                        `.metadata?['xxx']`
///   * map-literal writes: `metadata: { 'xxx': ..., 'yyy': ... }` — extracts
///                        every string-literal key inside the balanced `{}`
///   * `MetadataKeys.<ident>` references — resolved to the wire value via
///                        [constMap]
///
/// Does NOT cover:
///   * dynamic keys (`metadata[someVar]`) — by design
///   * non-literal nested access (`metadata['a']['b']`) — flagged as 'a' only
Set<String> _extractMetadataKeys(String source, Map<String, String> constMap) {
  final keys = <String>{};

  // `MetadataKeys.activityType` -> `activity_type`.
  final constRef = RegExp(r'MetadataKeys\.([A-Za-z0-9_]+)');
  for (final m in constRef.allMatches(source)) {
    final value = constMap[m.group(1)];
    if (value != null) keys.add(value);
  }

  // Subscript pattern — accepts either quote, allows !/? after metadata.
  final subscript = RegExp(
    r"""(?<![A-Za-z0-9_])metadata\s*[!?]?\s*\[\s*(['"])([a-z_][a-z0-9_]*)\1\s*\]""",
  );
  for (final m in subscript.allMatches(source)) {
    keys.add(m.group(2)!);
  }

  // Map-literal pattern — `metadata: {` or `metadata = {` or `metadata: <T>{`.
  // After matching, walk balanced braces and extract every `'xxx':` at
  // depth 1.
  final mapStart = RegExp(r"""(?<![A-Za-z0-9_])metadata\s*[:=]\s*(?:<[^>]+>\s*)?\{""");
  for (final m in mapStart.allMatches(source)) {
    final start = m.end; // just past the `{`
    int depth = 1;
    int i = start;
    while (i < source.length && depth > 0) {
      final c = source[i];
      if (c == '{') depth++;
      if (c == '}') depth--;
      if (depth == 1) {
        final rest = source.substring(i);
        final entry = RegExp(r"""^\s*(['"])([a-z_][a-z0-9_]*)\1\s*:""").matchAsPrefix(rest);
        if (entry != null) {
          keys.add(entry.group(2)!);
          i += entry.end;
          continue;
        }
      }
      i++;
    }
  }

  return keys;
}

File: apps/watch_wear/android/app/src/test/kotlin/com/runapp/watchwear/MetadataRegistryTest.kt
package com.runapp.watchwear

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/// Guard-rail: every `runs.metadata` key referenced in the Wear OS sources
/// must be registered in `docs/backend/metadata.md`.
///
/// `runs.metadata` is a jsonb bag with no type-level protection. Cross-client
/// drift (this watch writes `activity_type`, web reads `activityType`) is the
/// exact failure mode the registry was created to prevent. The Dart twin of
/// this guard is `apps/mobile_android/test/metadata_registry_test.dart`; the
/// TypeScript and Swift twins live under `apps/web/` and `apps/watch_ios/`.
/// Each scans its own platform's sources against the one shared registry.
class MetadataRegistryTest {

    /// Keys the guard should ignore because the match is a real `metadata`
    /// identifier that is not `runs.metadata`. Keep empty until a genuine
    /// spurious match appears, and always record the reason beside the entry.
    private val exemptReferences = emptySet<String>()

    @Test
    fun `every runs metadata key in Kotlin source is registered in docs backend metadata md`() {
        val doc = findUp("docs/backend/metadata.md")
        assertTrue("could not locate docs/backend/metadata.md from ${File(".").absolutePath}", doc != null)
        val registry = parseRegistry(doc!!.readText())
        assertTrue("parsed an empty registry — the markdown table shape changed", registry.isNotEmpty())

        val root = listOf(
            "src/main/kotlin",
            "app/src/main/kotlin",
            "apps/watch_wear/android/app/src/main/kotlin",
        ).firstNotNullOfOrNull { findUp(it) }
        assertTrue("could not locate the Wear OS main source set", root != null)

        val referenced = mutableMapOf<String, MutableSet<String>>()
        root!!.walkTopDown()
            .filter { it.isFile && it.extension == "kt" }
            .filterNot { it.path.contains("/generated/") }
            .forEach { file ->
                for (key in extractMetadataKeys(stripComments(file.readText()))) {
                    referenced.getOrPut(key) { mutableSetOf() }.add(file.path)
                }
            }

        // A scan that finds nothing is a broken scan, not a clean tree: the
        // watch's own run-metadata builder is always in range.
        assertTrue(
            "found no runs.metadata key references at all — the scan is not reaching the sources",
            referenced.isNotEmpty(),
        )

        val unknown = referenced.keys.filterNot { it in registry || it in exemptReferences }.sorted()
        assertEquals(
            buildString {
                append("Unregistered runs.metadata key(s) referenced in Kotlin source:\n")
                for (key in unknown) {
                    append("  $key\n")
                    referenced[key]!!.sorted().forEach { append("      at $it\n") }
                }
                append("\nEither:\n")
                append("  1) Register the key in docs/backend/metadata.md (preferred) — snake_case, ")
                append("shape, writers, readers, public_runs safety.\n")
                append("  2) If the match is spurious (a `metadata` identifier that is not ")
                append("runs.metadata), add the key to exemptReferences in this test with a reason.\n\n")
                append("Drift in runs.metadata is invisible to the DB type system — this registry ")
                append("IS the coordination point across web, mobile, watch_wear, and watch_ios.")
            },
            emptyList<String>(),
            unknown,
        )
    }

    /// Nearest ancestor of the working directory containing [rel].
    private fun findUp(rel: String): File? {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, rel)
            if (candidate.exists()) return candidate
            dir = dir.parentFile
        }
        return null
    }

    /// Registered key names from the markdown registry: each table row opens
    /// `| \`key\` |`. Same parse as the Dart guard, so the two can't disagree
    /// about what "registered" means.
    private fun parseRegistry(doc: String): Set<String> =
        Regex("""^\|\s*`([a-z_][a-z0-9_]*)`\s*\|""", RegexOption.MULTILINE)
            .findAll(doc)
            .map { it.groupValues[1] }
            .toSet()

    /// Blank `//` and block comments, leaving offsets intact. String literals
    /// are walked over rather than blanked, so a `"//"` inside a string can't
    /// be mistaken for a comment and swallow the rest of the line.
    private fun stripComments(source: String): String {
        val out = source.toCharArray()
        var i = 0
        while (i < source.length) {
            when {
                source[i] == '"' -> {
                    i++
                    while (i < source.length) {
                        if (source[i] == '\\') i++ else if (source[i] == '"') break
                        i++
                    }
                    i++
                }
                source.startsWith("//", i) -> {
                    while (i < source.length && source[i] != '\n') out[i++] = ' '
                }
                source.startsWith("/*", i) -> {
                    val end = source.indexOf("*/", i + 2)
                    val stop = if (end == -1) source.length else end + 2
                    while (i < stop) {
                        if (out[i] != '\n') out[i] = ' '
                        i++
                    }
                }
                else -> i++
            }
        }
        return String(out)
    }

    /// Every `runs.metadata` key reference in one comment-stripped file:
    ///
    ///   * subscript reads: `metadata["x"]`, `metadata!!["x"]`, `metadata?.get("x")`
    ///   * builder bodies: an identifier ending in `metadata` (`buildRunMetadata`,
    ///     `val metadata`) followed by `buildJsonObject { … }`, whose DEPTH-1
    ///     `put("x", …)` calls are the keys. Depth matters: the nested
    ///     `buildJsonArray`/`addJsonObject` inside `put("laps", …)` carries the
    ///     per-lap field names, which belong to the `laps` shape and are not
    ///     metadata keys of their own.
    private fun extractMetadataKeys(source: String): Set<String> {
        val keys = mutableSetOf<String>()

        SUBSCRIPT.findAll(source).forEach { keys.add(it.groupValues[1]) }

        for (m in BUILD_JSON_OBJECT.findAll(source)) {
            // Walk back over the declaration that owns this builder. It is a
            // metadata builder when the nearest identifier behind it ends in
            // `metadata` with no brace in between (so `fun buildRunMetadata(
            // … ): JsonObject = buildJsonObject {`, spanning lines, matches,
            // while an unrelated builder further down the file does not).
            val from = maxOf(0, m.range.first - DECLARATION_WINDOW)
            val window = source.substring(from, m.range.first)
            val name = METADATA_IDENTIFIER.findAll(window).lastOrNull() ?: continue
            val between = window.substring(name.range.last + 1)
            if (between.contains('{') || between.contains('}')) continue
            keys.addAll(depthOnePutKeys(source, m.range.last))
        }
        return keys
    }

    /// Keys of the `put("x", …)` calls at depth 1 of the brace block opened at
    /// [open].
    private fun depthOnePutKeys(source: String, open: Int): Set<String> {
        val keys = mutableSetOf<String>()
        var depth = 1
        var i = open + 1
        while (i < source.length && depth > 0) {
            when (source[i]) {
                '{' -> depth++
                '}' -> depth--
                '"' -> {
                    i++
                    while (i < source.length) {
                        if (source[i] == '\\') i++ else if (source[i] == '"') break
                        i++
                    }
                }
            }
            if (depth == 1) {
                PUT_CALL.matchAt(source, i)?.let { keys.add(it.groupValues[1]) }
            }
            i++
        }
        return keys
    }

    private companion object {
        const val DECLARATION_WINDOW = 400

        val SUBSCRIPT = Regex(
            """(?<![A-Za-z0-9_])[A-Za-z0-9_]*[Mm]etadata\s*(?:!!|\?)?\s*(?:\[|\.\s*get(?:Value)?\s*\()\s*"([a-z_][a-z0-9_]*)"""",
        )
        val BUILD_JSON_OBJECT = Regex("""buildJsonObject\s*\{""")
        val METADATA_IDENTIFIER = Regex("""[A-Za-z0-9_]*[Mm]etadata\b""")
        val PUT_CALL = Regex("""put\s*\(\s*"([a-z_][a-z0-9_]*)"""")
    }
}
